File: sdks/go/tests/torm_test.go
package torm_test

import (
	"os"
	"testing"

	"github.com/toonstore/torm-go"
)

var (
	testClient *torm.Client
	testURL    string
)

func TestMain(m *testing.M) {
	testURL = os.Getenv("TORM_URL")
	if testURL == "" {
		testURL = "http://localhost:3001"
	}
	testClient = torm.NewClient(testURL)
	os.Exit(m.Run())
}

// TestUser is a test model
type TestUser struct {
	ID      string `json:"id"`
	Name    string `json:"name"`
	Email   string `json:"email"`
	Age     int    `json:"age"`
	Website string `json:"website,omitempty"`
}

func (u *TestUser) GetID() string {
	return u.ID
}

func (u *TestUser) SetID(id string) {
	u.ID = id
}

func (u *TestUser) ToMap() map[string]interface{} {
	m := map[string]interface{}{
		"id":    u.ID,
		"name":  u.Name,
		"email": u.Email,
		"age":   u.Age,
	}
	if u.Website != "" {
		m["website"] = u.Website
	}
	return m
}

// TestProduct is a test model
type TestProduct struct {
	ID    string  `json:"id"`
	Name  string  `json:"name"`
	Price float64 `json:"price"`
	Stock int     `json:"stock"`
	SKU   string  `json:"sku"`
}

func (p *TestProduct) GetID() string {
	return p.ID
}

func (p *TestProduct) SetID(id string) {
	p.ID = id
}

func (p *TestProduct) ToMap() map[string]interface{} {
	return map[string]interface{}{
		"id":    p.ID,
		"name":  p.Name,
		"price": p.Price,
		"stock": p.Stock,
		"sku":   p.SKU,
	}
}

func TestClientCreation(t *testing.T) {
	client := torm.NewClient(testURL)
	if client == nil {
		t.Fatal("Failed to create client")
	}
}

func TestCreateDocument(t *testing.T) {
	users := torm.NewCollection(testClient, "testusers", func() *TestUser { return &TestUser{} })

	user := &TestUser{
		ID:    "test:user:1",
		Name:  "Alice",
		Email: "alice@example.com",
		Age:   30,
	}

	created, err := users.Create(user)
	if err != nil {
		t.Fatalf("Failed to create user: %v", err)
	}

	if created.GetID() != "test:user:1" {
		t.Errorf("Expected ID test:user:1, got %s", created.GetID())
	}
	if created.Name != "Alice" {
		t.Errorf("Expected name Alice, got %s", created.Name)
	}
}

func TestFindByID(t *testing.T) {
	users := torm.NewCollection(testClient, "testusers", func() *TestUser { return &TestUser{} })

	// Create user first
	user := &TestUser{
		ID:    "test:user:2",
		Name:  "Bob",
		Email: "bob@example.com",
		Age:   25,
	}
	_, err := users.Create(user)
	if err != nil {
		t.Fatalf("Failed to create user: %v", err)
	}

	// Find by ID
	found, err := users.FindByID("test:user:2")
	if err != nil {
		t.Fatalf("Failed to find user: %v", err)
	}

	if found.Name != "Bob" {
		t.Errorf("Expected name Bob, got %s", found.Name)
	}
}

func TestFindAll(t *testing.T) {
	users := torm.NewCollection(testClient, "testusers", func() *TestUser { return &TestUser{} })

	// Create multiple users
	users.Create(&TestUser{ID: "test:user:3", Name: "Charlie", Email: "charlie@example.com", Age: 35})
	users.Create(&TestUser{ID: "test:user:4", Name: "Diana", Email: "diana@example.com", Age: 28})

	// Find all
	all, err := users.Find()
	if err != nil {
		t.Fatalf("Failed to find all users: %v", err)
	}

	if len(all) < 2 {
		t.Errorf("Expected at least 2 users, got %d", len(all))
	}
}

func TestUpdateDocument(t *testing.T) {
	users := torm.NewCollection(testClient, "testusers", func() *TestUser { return &TestUser{} })

	// Create user
	user := &TestUser{
		ID:    "test:user:5",
		Name:  "Eve",
		Email: "eve@example.com",
		Age:   30,
	}
	created, err := users.Create(user)
	if err != nil {
		t.Fatalf("Failed to create user: %v", err)
	}

	// Update user
	created.Age = 31
	updated, err := users.Update(created.GetID(), created)
	if err != nil {
		t.Fatalf("Failed to update user: %v", err)
	}

	if updated.Age != 31 {
		t.Errorf("Expected age 31, got %d", updated.Age)
	}
}

func TestDeleteDocument(t *testing.T) {
	users := torm.NewCollection(testClient, "testusers", func() *TestUser { return &TestUser{} })

	// Create user
	user := &TestUser{
		ID:    "test:user:6",
		Name:  "Frank",
		Email: "frank@example.com",
		Age:   40,
	}
	created, err := users.Create(user)
	if err != nil {
		t.Fatalf("Failed to create user: %v", err)
	}

	// Delete user
	err = users.Delete(created.GetID())
	if err != nil {
		t.Fatalf("Failed to delete user: %v", err)
	}

	// Verify deletion
	_, err = users.FindByID(created.GetID())
	if err == nil {
		t.Error("Expected error when finding deleted user, got nil")
	}
}

func TestQueryWithFilter(t *testing.T) {
	users := torm.NewCollection(testClient, "testusers", func() *TestUser { return &TestUser{} })

	// Create test data
	users.Create(&TestUser{ID: "test:user:7", Name: "George", Email: "george@example.com", Age: 25})
	users.Create(&TestUser{ID: "test:user:8", Name: "Hannah", Email: "hannah@example.com", Age: 35})

	// Query users older than 30
	query := map[string]interface{}{
		"filters": []map[string]interface{}{
			{
				"field":    "age",
				"operator": "gt",
				"value":    30,
			},
		},
	}

	results, err := users.Query(query)
	if err != nil {
		t.Fatalf("Failed to query users: %v", err)
	}

	if len(results) < 1 {
		t.Error("Expected at least 1 user with age > 30")
	}

	for _, user := range results {
		if user.Age <= 30 {
			t.Errorf("Expected age > 30, got %d", user.Age)
		}
	}
}

func TestCount(t *testing.T) {
	users := torm.NewCollection(testClient, "testusers", func() *TestUser { return &TestUser{} })

	count, err := users.Count()
	if err != nil {
		t.Fatalf("Failed to count users: %v", err)
	}

	if count < 0 {
		t.Errorf("Expected non-negative count, got %d", count)
	}
}

func TestProductModel(t *testing.T) {
	products := torm.NewCollection(testClient, "testproducts", func() *TestProduct { return &TestProduct{} })

	product := &TestProduct{
		ID:    "test:product:1",
		Name:  "Laptop",
		Price: 999.99,
		Stock: 10,
		SKU:   "LAP-12345",
	}

	created, err := products.Create(product)
	if err != nil {
		t.Fatalf("Failed to create product: %v", err)
	}

	if created.SKU != "LAP-12345" {
		t.Errorf("Expected SKU LAP-12345, got %s", created.SKU)
	}

	if created.Price != 999.99 {
		t.Errorf("Expected price 999.99, got %f", created.Price)
	}
}

File: sdks/nodejs/tests/torm.test.ts
/**
 * TORM Node.js SDK Tests
 * 
 * Tests for ToonStore ORM Client
 */

import { TormClient, Model } from '../src/index';

// Test configuration
const TEST_CONFIG = {
  host: process.env.REDIS_HOST || 'localhost',
  port: parseInt(process.env.REDIS_PORT || '6379'),
};

interface TestUser {
  name: string;
  email: string;
  age: number;
  website?: string;
}

interface TestProduct {
  name: string;
  price: number;
  stock: number;
  sku: string;
}

describe('TormClient', () => {
  let torm: TormClient;

  beforeAll(() => {
    torm = new TormClient(TEST_CONFIG);
  });

  afterAll(async () => {
    await torm.disconnect();
  });

  test('should connect to database', async () => {
    const health = await torm.health();
    expect(health.status).toBe('ok');
    expect(health.connected).toBe(true);
  });

  test('should check connection status', () => {
    expect(torm.isConnected()).toBe(true);
  });

  test('should create a model', () => {
    const User = torm.model<TestUser>('TestUser');
    expect(User).toBeInstanceOf(Model);
  });
});

describe('Model CRUD Operations', () => {
  let torm: TormClient;
  let User: Model<TestUser>;

  beforeAll(() => {
    torm = new TormClient(TEST_CONFIG);
    User = torm.model<TestUser>('TestUser', {
      name: { type: 'string', required: true, minLength: 3 },
      email: { type: 'string', required: true, email: true },
      age: { type: 'number', required: true, min: 13, max: 120 },
      website: { type: 'string', url: true },
    });
  });

  afterAll(async () => {
    await User.deleteMany();
    await torm.disconnect();
  });

  beforeEach(async () => {
    await User.deleteMany();
  });

  test('should create a document', async () => {
    const user = await User.create({
      name: 'Alice',
      email: 'alice@example.com',
      age: 30,
    });

    expect(user._id).toBeDefined();
    expect(user.name).toBe('Alice');
    expect(user.email).toBe('alice@example.com');
    expect(user.age).toBe(30);
    expect(user._createdAt).toBeDefined();
    expect(user._updatedAt).toBeDefined();
  });

  test('should find all documents', async () => {
    await User.create({ name: 'Alice', email: 'alice@example.com', age: 30 });
    await User.create({ name: 'Bob', email: 'bob@example.com', age: 25 });

    const users = await User.find();
    expect(users.length).toBe(2);
  });

  test('should find document by ID', async () => {
    const created = await User.create({
      name: 'Alice',
      email: 'alice@example.com',
      age: 30,
    });

    const found = await User.findById(created._id);
    expect(found).toBeDefined();
    expect(found?._id).toBe(created._id);
    expect(found?.name).toBe('Alice');
  });

  test('should find one document by filter', async () => {
    await User.create({ name: 'Alice', email: 'alice@example.com', age: 30 });
    await User.create({ name: 'Bob', email: 'bob@example.com', age: 25 });

    const found = await User.findOne({ email: 'bob@example.com' });
    expect(found).toBeDefined();
    expect(found?.name).toBe('Bob');
  });

  test('should update a document', async () => {
    const user = await User.create({
      name: 'Alice',
      email: 'alice@example.com',
      age: 30,
    });

    const updated = await User.update(user._id, { age: 31 });
    expect(updated).toBeDefined();
    expect(updated?.age).toBe(31);
    expect(updated?._updatedAt).not.toBe(user._updatedAt);
  });

  test('should delete a document', async () => {
    const user = await User.create({
      name: 'Alice',
      email: 'alice@example.com',
      age: 30,
    });

    const deleted = await User.delete(user._id);
    expect(deleted).toBe(true);

    const found = await User.findById(user._id);
    expect(found).toBeNull();
  });

  test('should count documents', async () => {
    await User.create({ name: 'Alice', email: 'alice@example.com', age: 30 });
    await User.create({ name: 'Bob', email: 'bob@example.com', age: 25 });

    const count = await User.count();
    expect(count).toBe(2);
  });

  test('should delete all documents', async () => {
    await User.create({ name: 'Alice', email: 'alice@example.com', age: 30 });
    await User.create({ name: 'Bob', email: 'bob@example.com', age: 25 });

    const deleted = await User.deleteMany();
    expect(deleted).toBe(2);

    const count = await User.count();
    expect(count).toBe(0);
  });

  test('should delete documents matching filter', async () => {
    await User.create({ name: 'Alice', email: 'alice@example.com', age: 30 });
    await User.create({ name: 'Bob', email: 'bob@example.com', age: 25 });
    await User.create({ name: 'Charlie', email: 'charlie@example.com', age: 25 });

    const deleted = await User.deleteMany({ age: 25 });
    expect(deleted).toBe(2);

    const remaining = await User.find();
    expect(remaining.length).toBe(1);
    expect(remaining[0].name).toBe('Alice');
  });
});

describe('Validation', () => {
  let torm: TormClient;
  let User: Model<TestUser>;

  beforeAll(() => {
    torm = new TormClient(TEST_CONFIG);
    User = torm.model<TestUser>('TestUser', {
      name: { type: 'string', required: true, minLength: 3, maxLength: 50 },
      email: { type: 'string', required: true, email: true },
      age: { type: 'number', required: true, min: 13, max: 120 },
      website: { type: 'string', url: true },
    });
  });

  afterAll(async () => {
    await User.deleteMany();
    await torm.disconnect();
  });

  beforeEach(async () => {
    await User.deleteMany();
  });

  test('should validate required fields', async () => {
    await expect(
      User.create({ name: 'Alice', age: 30 } as any)
    ).rejects.toThrow('required');
  });

  test('should validate string minLength', async () => {
    await expect(
      User.create({ name: 'Al', email: 'alice@example.com', age: 30 })
    ).rejects.toThrow('at least 3 characters');
  });

  test('should validate string maxLength', async () => {
    await expect(
      User.create({
        name: 'A'.repeat(51),
        email: 'alice@example.com',
        age: 30,
      })
    ).rejects.toThrow('at most 50 characters');
  });

  test('should validate email format', async () => {
    await expect(
      User.create({ name: 'Alice', email: 'invalid-email', age: 30 })
    ).rejects.toThrow('valid email');
  });

  test('should validate URL format', async () => {
    await expect(
      User.create({
        name: 'Alice',
        email: 'alice@example.com',
        age: 30,
        website: 'not-a-url',
      })
    ).rejects.toThrow('valid URL');
  });

  test('should validate number min', async () => {
    await expect(
      User.create({ name: 'Alice', email: 'alice@example.com', age: 12 })
    ).rejects.toThrow('at least 13');
  });

  test('should validate number max', async () => {
    await expect(
      User.create({ name: 'Alice', email: 'alice@example.com', age: 121 })
    ).rejects.toThrow('at most 120');
  });

  test('should validate type', async () => {
    await expect(
      User.create({ name: 'Alice', email: 'alice@example.com', age: '30' } as any)
    ).rejects.toThrow('type number');
  });
});

describe('Query Builder', () => {
  let torm: TormClient;
  let User: Model<TestUser>;

  beforeAll(async () => {
    torm = new TormClient(TEST_CONFIG);
    User = torm.model<TestUser>('TestUser');

    await User.deleteMany();

    // Create test data
    await User.create({ name: 'Alice', email: 'alice@example.com', age: 30 });
    await User.create({ name: 'Bob', email: 'bob@example.com', age: 25 });
    await User.create({ name: 'Charlie', email: 'charlie@example.com', age: 35 });
    await User.create({ name: 'Diana', email: 'diana@example.com', age: 28 });
  });

  afterAll(async () => {
    await User.deleteMany();
    await torm.disconnect();
  });

  test('should filter with equals', async () => {
    const results = await User.query().where('age', 'eq', 30).exec();
    expect(results.length).toBe(1);
    expect(results[0].name).toBe('Alice');
  });

  test('should filter with greater than', async () => {
    const results = await User.query().where('age', 'gt', 30).exec();
    expect(results.length).toBe(1);
    expect(results[0].name).toBe('Charlie');
  });

  test('should filter with greater than or equal', async () => {
    const results = await User.query().where('age', 'gte', 30).exec();
    expect(results.length).toBe(2);
  });

  test('should filter with less than', async () => {
    const results = await User.query().where('age', 'lt', 30).exec();
    expect(results.length).toBe(2);
  });

  test('should filter with less than or equal', async () => {
    const results = await User.query().where('age', 'lte', 30).exec();
    expect(results.length).toBe(3);
  });

  test('should filter with contains', async () => {
    const results = await User.query()
      .where('email', 'contains', 'alice')
      .exec();
    expect(results.length).toBe(1);
    expect(results[0].name).toBe('Alice');
  });

  test('should chain multiple filters', async () => {
    const results = await User.query()
      .where('age', 'gte', 25)
      .where('age', 'lte', 30)
      .exec();
    expect(results.length).toBe(3);
  });

  test('should sort ascending', async () => {
    const results = await User.query().sort('age', 'asc').exec();
    expect(results[0].name).toBe('Bob');
    expect(results[3].name).toBe('Charlie');
  });

  test('should sort descending', async () => {
    const results = await User.query().sort('age', 'desc').exec();
    expect(results[0].name).toBe('Charlie');
    expect(results[3].name).toBe('Bob');
  });

  test('should limit results', async () => {
    const results = await User.query().limit(2).exec();
    expect(results.length).toBe(2);
  });

  test('should skip results', async () => {
    const results = await User.query()
      .sort('age', 'asc')
      .skip(2)
      .exec();
    expect(results.length).toBe(2);
    expect(results[0].age).toBeGreaterThanOrEqual(30);
  });

  test('should combine filter, sort, limit, and skip', async () => {
    const results = await User.query()
      .where('age', 'gte', 25)
      .sort('age', 'asc')
      .skip(1)
      .limit(2)
      .exec();
    expect(results.length).toBe(2);
    expect(results[0].age).toBe(28);
  });

  test('should count filtered results', async () => {
    const count = await User.query().where('age', 'gte', 30).count();
    expect(count).toBe(2);
  });

  test('should use equals shorthand', async () => {
    const results = await User.query().equals('name', 'Bob').exec();
    expect(results.length).toBe(1);
    expect(results[0].email).toBe('bob@example.com');
  });
});

describe('Custom Validation', () => {
  let torm: TormClient;
  let Product: Model<TestProduct>;

  beforeAll(() => {
    torm = new TormClient(TEST_CONFIG);
    Product = torm.model<TestProduct>('TestProduct', {
      name: { type: 'string', required: true },
      price: {
        type: 'number',
        required: true,
        min: 0,
        validate: (value) => value > 0,
      },
      stock: { type: 'number', required: true, min: 0 },
      sku: {
        type: 'string',
        required: true,
        pattern: /^[A-Z]{3}-\d{5}$/,
      },
    });
  });

  afterAll(async () => {
    await Product.deleteMany();
    await torm.disconnect();
  });

  beforeEach(async () => {
    await Product.deleteMany();
  });

  test('should validate with pattern', async () => {
    const product = await Product.create({
      name: 'Laptop',
      price: 999.99,
      stock: 10,
      sku: 'LAP-12345',
    });
    expect(product.sku).toBe('LAP-12345');
  });

  test('should reject invalid pattern', async () => {
    await expect(
      Product.create({
        name: 'Laptop',
        price: 999.99,
        stock: 10,
        sku: 'invalid',
      })
    ).rejects.toThrow('pattern');
  });

  test('should validate with custom function', async () => {
    await expect(
      Product.create({
        name: 'Laptop',
        price: 0,
        stock: 10,
        sku: 'LAP-12345',
      })
    ).rejects.toThrow('custom validation');
  });
});

console.log('✅ All TORM tests defined');
